llumnix/global_scheduler/migration_filter.py

```diff
@@ -0,0 +1,146 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, Dict, List, Optional
+from abc import ABC, abstractmethod
+
+from llumnix.logger import init_logger
+from llumnix.instance_info import InstanceInfo
+from llumnix.global_scheduler.scaling_scheduler import InstanceType
+from llumnix.global_scheduler.migration_policy import PairMigrationConstraints
+
+logger = init_logger(__name__)
+
+class MigrationFilterConfig:
+    def __init__(self, migrate_out_load_threshold):
+        self.migrate_out_load_threshold: float = migrate_out_load_threshold
+
+# TODO(KuilongCui): A filter might contain other filters; leave this for the future.
+class MigrationFilterPolicy(ABC):
+    @abstractmethod
+    def filter_src_condition(self, filter_config, pair_migration_type) -> Callable[[InstanceInfo], bool]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def filter_dst_condition(self, filter_config, pair_migration_type) -> Callable[[InstanceInfo], bool]:
+        raise NotImplementedError
+
+class MigrationFilter(ABC):
+    def __init__(self, filter_config: MigrationFilterConfig) -> None:
+        self.filter_config = filter_config
+        self.registered_filters: Dict[str, MigrationFilterPolicy] = {}
+
+    def register_filter(self, filter_name: str, migration_filter: MigrationFilterPolicy) -> bool:
+        if filter_name in self.registered_filters:
+            logger.warning("migration filter {} has been registered.".format(filter_name))
+            return False
+
+        self.registered_filters[filter_name] = migration_filter
+        return True
+
+    def unregister_filter(self, filter_name: str) -> None:
+        self.registered_filters.pop(filter_name, None)
+
+    def filter_instances(self, instance_infos: List[InstanceInfo],
+                         pair_migration_type: PairMigrationConstraints) -> Dict[str, InstanceInfo]:
+        src_filter_conditions = [filter.filter_src_condition() for filter in self.registered_filters.values()]
+        dst_filter_conditions = [filter.filter_dst_condition() for filter in self.registered_filters.values()]
+
+        if pair_migration_type == PairMigrationConstraints.NO_CONSTRAINTS:
+            policy_filter = MigrationFilterPolicyFactory.get_policy("loadconstraint")
+        elif pair_migration_type in [PairMigrationConstraints.PREFILL_2_DECODING, PairMigrationConstraints.DECODING_2_DECODING]:
+            policy_filter = MigrationFilterPolicyFactory.get_policy('pdd')
+        else:
+            raise ValueError(f"Unsupported pair migration type: {pair_migration_type}")
+        src_filter_conditions.append(policy_filter.filter_src_condition(self.filter_config, pair_migration_type))
+        dst_filter_conditions.append(policy_filter.filter_dst_condition(self.filter_config, pair_migration_type))
```

Comment:
I'm a bit confused. In the PDD design, lines 59-66 are able to filter out prefill/decoding instances, but lines 56-57 can also do the same thing, since `registered_filters` can also select `PddFilter`.

Reply:
registered_filters is used to register some additional filters. For example, instances that cannot participate in migration (perhaps not yet added to the current migration communication group), or to blacklist certain nodes, preventing them from being used as migration targets.

Reply:
registered_filters is used to register some additional filters. For example, instances that cannot participate in migration (perhaps not yet added to the current migration communication group), or to blacklist certain nodes, preventing them from being used as migration targets.

For line 59-66，the filter is used for general migration.